php/class-wp-seo.php

```diff
@@ -593,7 +674,232 @@ public function wp_head() {
+				$key = 'archive_date';
+			}
+
+			if ( empty( $key ) ) {
+				return $robots;
+			}
+
+			$robots = $this->filter_robots_meta( $robots, $key );
+
+			return $robots;
+		}
+
+		/**
+		 * Get the post meta or term option for the current object.
+		 * 
+		 * @return array Associative array of meta/option values, or empty array if none exist.
+		 */
+		public function get_current_object_options(): array {
+			// Singular posts.
+			if ( is_singular() ) {
+				$post_id = get_queried_object_id();
+				if ( empty( $post_id ) ) {
+					return [];
+				}
+
+				$post_meta = get_post_meta( $post_id );
+
+				return ! empty( $post_meta ) && is_array( $post_meta )
+					? $post_meta
+					: [];
+
+			}
+
+			// Term archives.
+			if ( is_category() || is_tag() || is_tax() ) {
+				$term = get_queried_object();
+				if ( empty( $term ) || ! $term instanceof WP_Term ) {
+					return [];
+				}
+
+				$term_options = get_option( $this->get_term_option_name( $term ) );
+
+				return ! empty( $term_options ) && is_array( $term_options )
+					? $term_options
+					: [];
+			}
+
+			return [];
+		}
+
+		/**
+		 * Set the robots meta directives based on current object
+		 * options (post meta or term options), or settings.
+		 * 
+		 * @param array  $robots Associative array of directives.
+		 * @param string $key    The settings key for the current page type.
+		 * @return array Updated associative array of directives.
+		 */
+		public function filter_robots_meta( array $robots, string $key ): array {
+			if ( empty( $key ) ) {
+				return $robots;
+			}
+
+			// List of possible robots directives.
+			$robots_directives = $this->get_robots_directive_values();
+
+			if ( empty( $robots_directives ) || ! is_array( $robots_directives ) ) {
+				return $robots;
+			}
+
+			// Get the data for the current object (post meta or term options), if any.
+			$object_data = $this->get_current_object_options();
+
+			// Get the settings from the settings page.
+			$settings = WP_SEO_Settings()->get_option( "{$key}_robots" );
+
+			// Check the meta value for each directive and set it to true if present.
+			foreach ( $robots_directives as $directive ) {
+				if ( empty( $directive ) ) {
+					continue;
+				}
+
+				$robots_meta = '';
+
+				if ( is_singular() && ! empty( $object_data ) ) {
+					$robots_meta = $object_data[ "_meta_robots_{$directive}" ][0] ?? '';
+				} elseif ( ( is_category() || is_tag() || is_tax() ) && ! empty( $object_data ) ) {
+					$robots_meta = $object_data[ "robots_{$directive}" ] ?? '';
+				}
+
+				if ( empty( $robots_meta ) && ! empty( $settings ) && is_array( $settings ) ) {
+					/**
+					 * Filter the robots meta directive for this page.
+					 * @param  string      The value retrieved from the settings.
+					 * @param  string $key The key of the setting retrieved.
+					 */
+					$robots_meta = apply_filters(
+						"wp_seo_robots_{$directive}_value",
+						in_array( $directive, $settings ) ? $directive : '',
```

Comment:
```suggestion
						in_array( $directive, $settings ) ? $settings[ $directive ] : '',
```
I think I'm understanding this right?

Reply:
maybe I'm not... if so, ignore me.

Reply:
`$settings` is actually a simple array of directives (e.g. `[ 'noindex' ]`) – I updated the logic so it's more clear!